lib/segment/batcher.ex

```diff
@@ -0,0 +1,74 @@
+defmodule Segment.Analytics.Batcher do
+  @moduledoc """
+
+  """
+  use GenServer
+  alias Segment.Analytics.{Track, Identify, Screen, Alias, Group, Page}
+
+  @max_batch_size Application.get_env(:segment, :max_batch_size, 100)
+  @batch_every_ms Application.get_env(:segment, :batch_every_ms, 2000)
+
+  def start_link(api_key) do
+    client = Segment.Http.client(api_key)
+    GenServer.start_link(__MODULE__, {client, :queue.new()}, name: __MODULE__)
+  end
+
+  def start_link(api_key, adapter) do
+    client = Segment.Http.client(api_key, adapter)
+    GenServer.start_link(__MODULE__, {client, :queue.new()}, name: __MODULE__)
+  end
+
+  # client
+  def call(event = %Track{}), do: enqueue(event)
+  def call(event = %Identify{}), do: enqueue(event)
+  def call(event = %Screen{}), do: enqueue(event)
+  def call(event = %Alias{}), do: enqueue(event)
+  def call(event = %Group{}), do: enqueue(event)
+  def call(event = %Page{}), do: enqueue(event)
+
+  # GenServer Callbacks
+
+  @impl true
+  def init({client, queue}) do
+    schedule_batch_send()
+    {:ok, {client, queue}}
+  end
+
+  @impl true
+  def handle_cast({:enqueue, event}, {client, queue}) do
+    {:noreply, {client, :queue.in(event, queue)}}
+  end
+
+  @impl true
+  def handle_info(:process_batch, {client, queue}) do
+    length = :queue.len(queue)
+    {items, queue} = extract_batch(queue, length)
+
+    Segment.Http.batch(client, items)
```

Comment:
Should we guard against empty batches here being sent? Didn't see that in the http sender

Reply:
would this work `Segment.Http.batch(client, items) when length(items) > 0` ?

Reply:
```    case items do
      [] ->
        :ok

      items ->
        Segment.Http.batch(client, items)
    end```

Reply:
or is `if length(items) > 0, do: Segment.Http.batch(client, items)` preferable

Reply:
`if` case fine here I think